### ISSUE TITLE
Graceful shutdown of producers - general question

### ISSUE THREAD

**Issue by pawelrychlik** [tulios/kafkajs#879]
Hi 👋,

I have this sort of a general question here.

We've been using kafka for a few months now, although so far mostly for non-critical processing. Now, we'd like to take it to the next level and ideally end up with a setup that can be 100% relied on, especially when it comes to producing messages.

In the past months, we've experienced quite a lot of random broker failures, disconnects, few seconds/minutes downtimes, triggering leader election, consumer group rebalance, and otherwise some intermittent hiccups. I think the issue has been mostly solved by now (thanks to AWS MSK support tweaks on our kafka clusters), but still we need high availability, we cannot afford loosing messages.

So as a solution to ensure reliability in message production, we started to play around with the kafkajs retry parameters. In theory, we could take it to the extreme with only-once semantics, idempotent mode on, and infinite retries.

But we bumped into this problem with kafkajs: **`await producer.disconnect()`** (the command we execute on SIGINT in our app shutdown procedure) **doesn't actually wait till all the in-memory messages that are currently being retried by the producer get sent out**.

I'd like to have a solution where:
a) either - `producer.disconnect()` waits until it successfully sends out all outstanding messages (or KafkaJSNumberOfRetriesExceeded error is thrown, whichever comes first).
b) or - `producer.disconnect()` causes the messages that are currently in-flight (i.e. being retried) to throw an error, so that we can apply some fall-back logic for saving those messages elsewhere just to avoid loosing them irreversibly. 🤷‍♂️ 

Sure, we could implement some custom wrapper on top of kafkajs, with message counters, timers etc, just to know if it's still retrying or if it's all done, and can be shut down safely. But it feels like overengineering to me.

The ultimate question: How do you ensure graceful shutdown of kafkajs producers, in a way that will never cause loosing messages?

Thanks in advance for any recommendations,
Pawel

**Comment by Nevon** [comment]
I believe this was actually fixed in #869. As you mentioned, we used to just discard any messages currently in the request queue on disconnect, but this has now changed so that we first wait for the queue to be drained. You can try it out in the latest beta version and see if it solves your issue.

**Comment by tulios** [comment]
1.14 was already released with the fix.

**Comment by pawelrychlik** [comment]
Hi again,

First of all, thanks for your quick replies. I really appreciate it. 🙇

I didn't have enough time to dig into this issue myself in the past months, other than just do a quick check with the versions you mentioned. I've tried with `1.13.0-beta.62`. Then with `1.14.0` once it was released. Now, as there's been a brand new release recently, I tried with `1.15.0`. I didn't want to bother you needlessly without proper argumentation at hand. But I think I do have it now :)

I'm working with 1.15.0 now. I git-cloned this repo. I'm running this [example producer](https://github.com/tulios/kafkajs/blob/master/examples/producer.js) against my own docker-compose-hosted kafka (`wurstmeister/kafka:2.12-2.4.0`, with SSL disabled for simplicity and [these lines](https://github.com/tulios/kafkajs/blob/master/examples/producer.js#L14-L23) commented out, but I guess it doesn't matter).

1. The producer is sending messages to kafka just fine 👌.
2. I kill my docker-compose kafka, the script starts to notice it after a few seconds - I see first errors coming in 👌.
3. I signal a SIGINT (Ctrl+C) to the producer script. The process continues to run and throw errors at me - which I suppose is thanks to the fix you mentioned 👌.
_(Digression: it's a bit weird, but the producer continues to run long after the SIGINT, even despite my kafka is up already. I believe it should shut down by itself, once all pending retries either succeed or fail. Perhaps some dangling timeout/interval handle in node.js? I didn't analyse it)._
4. I Ctrl+C again and the process gets killed. 👌

So yeah, the first impression is indeed that the producer now waits till all messages are sent out or fail.

But that still doesn't work as expected in my project, so I started to experiment. I've modified [the example producer](https://github.com/tulios/kafkajs/blob/master/examples/producer.js) in a way that I've added an extra `kafka.logger().info('[example/producer] disconnected')` right after [this line](https://github.com/tulios/kafkajs/blob/master/examples/producer.js#L92). This way I know exactly when `await producer.disconnect()` is completed.

To my surprise, the `disconnect()` happens immediately, as if there was nothing to `await` on. Logs:

```
// kafka is down, so these errors are expected:
// [...]
error: ┏ [BrokerPool] Failed to connect to seed broker, trying another broker from the list: Connection error: connect ECONNREFUSED 127.0.0.1:9094
error: ┃ [0] {
error: ┃ [1]   "timestamp": "2020-12-01T13:36:04.150Z",
error: ┃ [2]   "logger": "kafkajs",
error: ┃ [3]   "retryCount": 2,
error: ┃ [4]   "retryTime": 1224
error: ┗ [5] }
// here I signal a SIGINT:
^C
info:  ┏ [example/producer] disconnecting on SIGINT
info:  ┃ [0] {
info:  ┃ [1]   "timestamp": "2020-12-01T13:36:04.397Z",
info:  ┃ [2]   "logger": "kafkajs"
info:  ┗ [3] }
// here goes my custom log:
info:  ┏ [example/producer] disconnected
info:  ┃ [0] {
info:  ┃ [1]   "timestamp": "2020-12-01T13:36:04.397Z",
info:  ┃ [2]   "logger": "kafkajs"
info:  ┗ [3] }
// errors continue:
error: ┏ [Connection] Connection error: connect ECONNREFUSED 127.0.0.1:9094
error: ┃ [0] {
error: ┃ [1]   "timestamp": "2020-12-01T13:36:05.210Z",
error: ┃ [2]   "logger": "kafkajs",
error: ┃ [3]   "broker": "localhost:9094",
error: ┃ [4]   "clientId": "example-producer",
error: ┃ [5]   "stack": "Error: connect ECONNREFUSED 127.0.0.1:9094\n    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1113:14)"
error: ┗ [6] }
// [...]
```

Bottom line: awaiting on `producer.disconnect()` clearly doesn't have the expected effect.

What do you recommend?

/Thanks in advance.
Pawel

**Comment by ericmorand** [comment]
I have the same issue here: calling `producer.disconnect()` doesn't guarantee that the promise is resolved when all the pending messages have been sent.

@Nevon , @tulios , is there something we are missing here?
